rotkehlchen/chain/ethereum/modules/makerdao/sai/decoder.py

```diff
@@ -0,0 +1,505 @@
+    Information relating to event topics and contracts came from:
+    https://github.com/makerdao/sai/blob/d28fe5074af7c661fb3aa7f117ad419fb6414b91/DEVELOPING.md
+    """
+    def __init__(
+            self,
+            ethereum_inquirer: 'EthereumInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(ethereum_inquirer, base_tools, msg_aggregator)
+        self.ethereum_inquirer = ethereum_inquirer
+        self.base_tools = base_tools
+        self.weth = A_WETH.resolve_to_evm_token()
+        self.eth = A_ETH.resolve_to_crypto_asset()
+        self.sai = A_SAI.resolve_to_evm_token()
+        # NotERC20Conformant is never raised because PETH is a valid ERC20 token
+        self.peth = get_or_create_evm_token(
+            userdb=self.base_tools.database,
+            chain_id=ChainID.ETHEREUM,
+            token_kind=EvmTokenKind.ERC20,
+            evm_address=POOLED_ETHER_ADDRESS,
+            evm_inquirer=self.ethereum_inquirer,
+        )
+
+    def _decode_sai_tub(
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],
+            action_items: Optional[list[ActionItem]],
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        try:
+            decoder_function = getattr(self, TOPICS_TO_METHODS[tx_log.topics[0]])
+            return decoder_function(
+                tx_log=tx_log,
+                transaction=transaction,
+                decoded_events=decoded_events,
+                all_logs=all_logs,
+                action_items=action_items,
+            )
+        except KeyError:
+            return None, []
+
+    def _decode_withdraw_eth(
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,  # pylint: disable=unused-argument
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[list[ActionItem]],  # pylint: disable=unused-argument
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        """
+        This method decodes ETH withdrawal from a CDP.
+
+        An example of such transaction is:
+        https://etherscan.io/tx/0x4e569aa1f23dc771f1c9ad05ab7cdb0af2607358b166a8137b702f81b88e37b9
+        """
+        cdp_id = hex_or_bytes_to_int(tx_log.topics[2])
+        amount_withdrawn_raw = hex_or_bytes_to_int(tx_log.topics[3])
+        amount_withdrawn = asset_normalized_value(amount=amount_withdrawn_raw, asset=self.eth)
+
+        for event in decoded_events:
+            if (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.balance.amount.is_close(amount_withdrawn, max_diff=MAX_DIFF_ALLOWED) and
```

Comment:
This is still terrible. I am dissapointed you don't see it. You are doing something that will not always work. It's very very bad code.

You are making a comparison that will work for this transaction, maybe few more, but the "is_close(amount)" you just pulled it out of your ass

PETH and ETH probably have some price ratio **at the time** of the transaction. You can't just guess it like you did. This is not gonna fly and you should have seen this on your own. 

So more questions:

1. Why do you need PETH? In the transaction you linked you don't care about it. It's internal mechanisms and do not concern the user.
![2022-12-14_16-28](https://user-images.githubusercontent.com/1658405/207637844-e72f619f-6389-454b-a9a2-c768025e5134.png)
The user deposits ETH and gets SAI. That's all we care about.
2. If you really need to know the amount of PETH it's there in the transaction. You don't need to guess it. It has a mint event.
![2022-12-14_16-25](https://user-images.githubusercontent.com/1658405/207638573-4856d98d-74e8-4a91-b987-4dd478120bde.png)